### PR TITLE
Query: Lift Subqueries when outer QM's select can be pushed down into…

### LINKED ISSUE #7831: Query: Optimize QueryOptimizer
`QueryModelOptimizer` derives from `SubQueryFromClauseFlattener` class which in turn derives from `QueryModelVisitorBase`.
- QMVBase class does not visit, QueryModels recursively, hence subQMs inside SelectClause etc are not visited by optimizer. We need ExpressionVisitor to apply QueryModelOptimizer recursively.
- Expand logic of lifting of subquery.
- Verify if we need to derive from `SubQueryFromClauseFlattener`, its code structure has a specific format which we can get rid of.

### PR DESCRIPTION
**Pull request by smitpatel**
… subqueryQM

Apply QueryOptimizer recursively on all QueryModels
Introduce SubQueryExpressionVisitor<IQueryModelVisitor> to apply QMV's recursively. @maumar 

Partially fix #7831

### CODE REVIEW

File: src/EFCore/Query/ExpressionVisitors/Internal/SubQueryExpressionVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        protected override Expression VisitSubQuery(SubQueryExpression expression)
+        {
+            TransformingVisitor.VisitQueryModel(expression.QueryModel);
+            return expression;
```
Comment (maumar) [inline src/EFCore/Query/ExpressionVisitors/Internal/SubQueryExpressionVisitor.cs]: nit: add newline before return

Comment (smitpatel) [inline src/EFCore/Query/ExpressionVisitors/Internal/SubQueryExpressionVisitor.cs]: thanks for reminding, it should be changed to base visitor actually.

Comment (smitpatel) [inline src/EFCore/Query/ExpressionVisitors/Internal/SubQueryExpressionVisitor.cs]: And base returns expression only. :trollface:

Comment (smitpatel) [inline src/EFCore/Query/ExpressionVisitors/Internal/SubQueryExpressionVisitor.cs]: @maumar - Decide what you want? `return Expression` or call base? 😛

Comment (maumar) [inline src/EFCore/Query/ExpressionVisitors/Internal/SubQueryExpressionVisitor.cs]: call base is "more correct" ;)

**Comment by smitpatel** [comment]
@maumar - Updated.
I will submit changes to ExpressionVisitorBase in separate PR.

**Pull request merged**